src/eval.c

```diff
@@ -767,62 +556,6 @@ usage: (defvar SYMBOL &optional INITVALUE DOCSTRING)  */)
   return sym;
 }
 
-DEFUN ("defconst", Fdefconst, Sdefconst, 2, UNEVALLED, 0,
-       doc: /* Define SYMBOL as a constant variable.
-This declares that neither programs nor users should ever change the
-value.  This constancy is not actually enforced by Emacs Lisp, but
-SYMBOL is marked as a special variable so that it is never lexically
-bound.
-
-The `defconst' form always sets the value of SYMBOL to the result of
-evalling INITVALUE.  If SYMBOL is buffer-local, its default value is
-what is set; buffer-local values are not affected.  If SYMBOL has a
-local binding, then this form sets the local binding's value.
-However, you should normally not make local bindings for variables
-defined with this form.
-
-The optional DOCSTRING specifies the variable's documentation string.
-usage: (defconst SYMBOL INITVALUE [DOCSTRING])  */)
-  (Lisp_Object args)
-{
-  Lisp_Object sym, tem;
-
-  sym = XCAR (args);
-  Lisp_Object docstring = Qnil;
-  if (!NILP (XCDR (XCDR (args))))
```

Comment:
I can't actually convince myself that this C is safe.

How do we know that `XCDR (args)` is a cons cell, so we can call `XCDR` again without crashing?

Reply:
We know that args is a cons cell because `defconst` is marked as 'unevalled'. So both the `XCAR` and first `XCDR` are safe. In fact, these were changed from calls to `Fcar` and `Fcdr` because the `XCAR` is faster. It is not clear to me that the second call to `XCDR` is similarly guaranteed to be safe.

Reply:
Yep, it was the second XCDR I was worried about. Still, not a problem after this PR! :)

Reply:
I played with this today, but I've failed to crash Emacs with it.

```
ELISP> (eval (cons 'defconst (cons 'x (cons 1 (cons 2 nil)))))
x
ELISP> (eval (cons 'defconst (cons 'x (cons 1 'not-a-cons))))
*** Eval error ***  Wrong type argument: listp, not-a-cons
```

I think because the arguments are unevalled, and we require at least two arguments, it's safe to call `XCDR(XCDR ...)`.